realtek: Add support for ZyXEL XGS1210-12 Switch

leo5593 opened the pull request:
The ZyXEL XGS1210-12 Switch is a 10 + 2 port multi-GBit switch:
- 8 x 10/100/1000BaseT Ethernet ports
- 2 x 10/100/1000/2500BaseT Ethernet ports
- 2 SFP+ cages (1G and 10G)

Hardware:
- RTL9302B SoC
- Macronix MX25L12833F (16MB flash)
- Nanja NT5CC64M16GP-1 (128MB DDR3 SDRAM)
- RTL8231 GPIO extender to control the port LEDs
- RTL8218D 8x Gigabit PHY
- RTL8226 2x 2.5Gigabit PHY
- SFP+ 2x 10GBit slot

At this commit, all ethernet ports work under OpenWRT, including the various NBaseT modes and the SFP+ are working with 10G and 1G module with i2c support for both ports, tested with:
- SFP FS SFP1G-LX-31
- SFP FS SFP-10GLR-31
- SFP Arpers SFP-10G-T-COM
- DAC Molex 10G
- DAC FS SFP-PC02 1G

EDIT: The support for these SFP have been moved to an other branch.

There was a previous request with [23.05](https://github.com/openwrt/openwrt/pull/17161) this request has the minimum changes in order to test with the hardware I have. There was also other adjustement for the [led](https://github.com/openwrt/openwrt/commit/005edf9428b5efe38d1da3d26878cf4787a53cf4) register and [debugfs](https://github.com/openwrt/openwrt/commit/9f19c6321b99580597314bdfb7f53891ed3f5062) for testing.

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,369 @@
+
+		/* i2c of the left SFP+ cage as seen from the front: port 11 */
+		i2c0_0: i2c@0 {
+			reg = <9>;
```
More of a question to all. What's the correct value here? Some dts (including mine) use `<0>`, your's uses `<9>`. Same for `i2c0_1` accordingly. Mine works also fine...

leo5593 replied:
This was based on other commits, I just assumed it needed to match the "sda-pin = <10>;", but may be this is not even required.

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,369 @@
+				read-only;
+			};
+			partition@100000 {
+				label = "jffs";
```
Probably keep the naming of the stock firmware layout? There the partitions are named `jffs2-cfg` and `jffs2-log`.

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,369 @@
+			partition@fe0000 {
+				label = "log";
+				reg = <0xfe0000 0x20000>;
+			};
```
```suggestion
			    read-only;
			};
```

This partition is used by the bootloader for a self-check, expecting a specific value there (see [here](https://www.svanheule.net/switches/zyxel_xgs1210_series#oem_flash_layout) for details). So I suggest making it read-only.

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,369 @@
+			reg = <0>;
+			label = "lan1";
+			phy-handle = <&phy0>;
+			phy-mode = "xgmii";
```
```suggestion
			phy-mode = "usxgmii";
```

This works on 6.6 but will break on 6.12. See [here](https://github.com/openwrt/openwrt/pull/18935/commits/5b960a2ae27f6635bb3003b67ec35e5e9c301fb8). This issue came up in the WIP PR #18935 and has been adressed there, see the commit message for details. Same for the other 1G ports.

jonasjelonek reviewed (changes requested):
I'm working on support for this device too for quite some time, based on work of @oliv3r and @janh (although for XGS1010-12; see my tree [here](https://github.com/jonasjelonek/openwrt/tree/xgs1210-12)). But haven't done a PR so far for several reasons.

Formally, your commits lack a proper title and message. Especially when adding support for a new device, the commit adding the support usually has a description about the device and some steps how to run/install OpenWrt on that device. See my commit in my tree or the various other examples in OpenWrt. Same applies to the commit title.

Regarding the "content" of your PR, it is in most parts equal to mine and to what others did already. See the review parts for what I suspect to be some issues.

EDIT: In general, also have a look at the realtek 6.12 PR #18935 by @plappermaul. I suggest you also test it with kernel 6.12 to make sure upfront it works properly after this has been merged. This way, we can avoid "fixing" this device with 6.12 afterwards.

jonasjelonek commented:
Perspectively, we should join the single device trees of XGS1010-12, XGS1210-12 and XGS1250-12 (as @oliv3r already did in his PR) to have common parts in a single `dtsi`. But I actually prefer to add support for all three devices at first and then joining them in another PR.

leo5593 commented:
Hello @jonasjelonek,

I had made a PR with the 23.05 branch, I had some time and I did the same mistake of doing the work on the 24.10, I merged the changes without the proper due diligence regarding the main branch.

It is hard to keep track of the ongoing work about a device, I did the same mistake as in December, not looking for progress on the device.

For the steps how to run/install, I had mention the relation to the XGS1250-12 and related procedure, on a previous commit and forget to add the information.

I agree to add the support for the 3 devices and then merge and clean the device trees.

I have added the modification based on you comments and also rebased on top of lasted main with "Signed-off-by" added to each commit.

I will try to merged these commits with kernel612 to check as suggested.

Sorry for being so newbie, it's my first time trying to contribute.

Neustradamus commented:
@leo5593: Thanks for this new PR!

jonasjelonek reviewed (changes requested):
Some stuff probably got lost during your rebase. So I'm marking them here again. And please don't add fix commits, just edit the commits to include the fixes. This keeps history clean.

Signed-off-by seems fine now however, commit title and message are still not accurate. Commit title should begin with `realtek:` because this is specific to the realtek target. The message should either describe the device and how to use OpenWrt on it, or you link to the information for XGS1250-12, since this is mostly the same.

I don't like that you include the other changes for extending the MAC/SFP capabilities. This also affects other `rtl930x` devices and is IMHO better to be discussed in a separate PR.

jonasjelonek commented on target/linux/realtek/image/rtl930x.mk:
```diff
@@ -36,3 +36,19 @@ define Device/zyxel_xgs1250-12
+define Device/zyxel_xgs1210-12
+  SOC := rtl9302
+  UIMAGE_MAGIC := 0x93001210
+  ZYXEL_VERS := ABWE
```
`ABWE` is version code of XGS1250-12, for XGS1210-12 this should be `ABTY`

oliv3r commented:
> Some stuff probably got lost during your rebase. So I'm marking them here again. And please don't add fix commits, just edit the commits to include the fixes. This keeps history clean.

IMO during a review doing 'git commit --fixup' is fine, and even helps with the review. Just make sure at some point to rebase --autosquash them :)

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,355 @@
+
+		led_set0 = <0x0a20 0x0b80>; // LED set 0: 1000Mbps,  10/100Mbps
+		led_set1 = <0x0a20 0x0b80 0x0b8b 0x0a0b>;  // LED set 1: 10Mbit, 100Mbit, 1000Mbit, 2.5Gbit
+		led_set2 = <0x0a20 0x0a01>; // LED set 2: 1000MBit, 10GBit
```
Off-topic for this PR: Could you quickly point to where you fixed this? Still haven't figured out what these values actually mean right now.

janh replied:
> Still haven't figured out what these values actually mean right now.

These are the raw values of the LED hardware registers. See here, including link to documentation: https://github.com/openwrt/openwrt/pull/17137#issuecomment-2779832295

Do the LEDs on the 2.5G ports actually work correctly with this config? On XGS1010-12 v1, different values are needed, and I would have expected that the XGS1210 hardware is identical in that aspect:
https://github.com/janh/openwrt/blob/5f2f7457e8d41a43d11b4fca6712f982a8e9b57e/target/linux/realtek/dts/rtl9302_zyxel_xgs1010-12.dts#L89-L95

hailfinger commented:
There are two hardware revisions of this switch:
- Zyxel XGS1210 Desktop Gigabit Smart Switch, 10x RJ-45, 2x SFP+ (XGS1210-12-ZZ0101F)
- Zyxel XGS1210 Desktop Gigabit Smart Switch, 10x RJ-45, 2x SFP+, V2 (XGS1210-12-ZZ0102F)

Which revision is covered by this patch?

plappermaul commented:
Thanks for providing this patch. And another thanks that it only adds the DTS. Too many complex PRs have been initiated to get this device integrated.

Tested initramfs and network connection works on Port 1-8. With that it is fine for me.

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,356 @@
+
+		/* i2c of the left SFP+ cage as seen from the front: port 11 */
+		i2c0_0: i2c@0 {
+			reg = <9>;
```
```suggestion
			reg = <0>;
```

Still not sure about that one. In the driver it is used as channel number. Since this doesn't seem to have a functional difference, I'd suggest to use `<0>` here.

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,356 @@
+
+		/* i2c of the right SFP+ cage as seen from the front: port 12 */
+		i2c0_1: i2c@1 {
+			reg = <10>;
```
```suggestion
			reg = <1>;
```
Same as above

jonasjelonek commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,356 @@
+		compatible = "realtek,rtl9300-leds";
+		active-low;
+
+		led_set0 = <0x0a20 0x0b80>; // LED set 0: 1000Mbps,  10/100Mbps
+		led_set1 = <0x0a20 0x0b80 0x0b8b 0x0a0b>;  // LED set 1: 10Mbit, 100Mbit, 1000Mbit, 2.5Gbit
+		led_set2 = <0x0a20 0x0a01>; // LED set 2: 1000MBit, 10GBit
```
Please adjust the LED values according to what @janh mentioned. The existing values do not really work properly with the 2.5G ports.

https://github.com/janh/openwrt/blob/5f2f7457e8d41a43d11b4fca6712f982a8e9b57e/target/linux/realtek/dts/rtl9302_zyxel_xgs1010-12.dts#L84-L101

jonasjelonek reviewed (changes requested):
Some small things. Please also fix some typos in your commit message. Then I'll be fine with this.

leo5593 commented:
> Thank you for your contribution.
> 
> One nitpick from my side. Please use the correct casing of Zyxel and OpenWrt everywhere. e.g. commit messages.

I always wonder about this detail, that's why I used the commit from the XGS1250-12 for reference.
New commit on the way.

leo5593 committed:
realtek: Add support for Zyxel XGS1210-12 Switch

The Zyxel XGS1210-12 Switch is a 10 + 2 port multi-GBit switch with
8 x 1000BaseT, 2 x 10/100/1000/2500BaseT Ethernet ports and
2 SFP+ module slot.

Hardware:
- RTL9302B SoC
- Macronix MX25L12833F (16MB flash)
- Nanja NT5CC64M16GP-1 (128MB DDR3 SDRAM)
- RTL8231 GPIO extender to control the port LEDs
- RTL8218D 8x Gigabit PHY
- RTL8226 2x 10m/100m/1/2.5 Gigabit PHY
- SFP+ 2x 10GBit slot

Power is supplied via a 12V 1.5A standard barrel connector. At the
right side behind the grid is UART serial connector. A Serial
header can be connected to from the outside of the switch trough
the airvents with a standard 2.54mm header.
Pins are from top to bottom Vcc(3.3V), TX, RX and GND. Serial
connection is via 115200 baud, 8N1.

A reset button is accessble through a hole in the front panel

At the time of this commit, all ethernet ports work under OpenWrt,
including the various NBaseT modes, SFP+ slots are supported with i2c bus.

Installation
--------------
* Connect serial as per the layout above. Connection parameters: 115200 8N1.
* Navigate to 'Management' in the OEM web interface and click on 'Firmware upgrade'
  to the left.
* Upload the OpenWrt initramfs image, and wait till the switch reboots.
* Connect to the device through serial and change the U-boot boot command.
  > fw_setenv bootcmd 'rtk network on; boota'
* Reboot, scp the sysupgrade image to /tmp, verify the checksum and flash it:
  > sysupgrade openwrt-realtek-rtl930x-Zyxel_xgs1210-12-squashfs-sysupgrade.bin
* Upon reboot, you have a functional OpenWrt installation. Leave the bootcmd
  value as is - without 'rtk network on' the switch will fail to initialise
  the network.

Debug
------------
* Connect serial as per the layout above. Connection parameters: 115200 8N1.
* A tftp server is requiered, tftpd-hpa works well.
* Power the device, at U-Boot start rapidly hit Esc key to stop autoboot
* Enable network:
  > rtk network on
* Change ip address (default is 192.168.1.1):
  > setenv ipaddr 192.168.1.6
* Download initramfs:
  > tftpboot 0x84f00000 192.168.1.111:openwrt-realtek-rtl930x-Zyxel_xgs1210-12-initramfs-kernel.bin
* Boot loaded file:
  > bootm 0x84f00000

This prodecudre also apply to the sock firmware with the file XGS1210-12_V2.00(ABTY.1)C0.bix.

More information can be found on the page of XGS1250-12 as they share the same base.

Signed-off-by: Nicolas BERTRAND <nicolasbertrand89@gmail.com>

jonasjelonek reviewed (approved):
I'm fine with this now, thanks!

oliv3r commented:
> > I've done several attempts, making the set smaller and smaller; but constantly ran into 'not perfect' where perfect was unatainable due to a) the horrible mess that is the realtek support so far, and b) stepping stones where needed to get to the perfect state as intermediate steps, to improve the mess, which was also rejected. Let alone 'get stuff working within this horrible mess.
> 
> Times have changed. Join back the party if you like.

You mean we can actually get patches merged and stuff moving again?!

I did actually consider buying a nother switch (netgear GS108v3) as RTL8xxx platform .... I just need some mental bandwidth, I'm extremly short on time right now, but the itch is certainly there still :D

GoetzGoerisch reviewed (commented):
LGTM thanks!

hailfinger commented:
> You mean we can actually get patches merged and stuff moving again?!

Indeed. And your past efforts are highly appreciated.

> I did actually consider buying a nother switch (netgear GS108v3) as RTL8xxx platform .... I just need some mental bandwidth, I'm extremly short on time right now, but the itch is certainly there still :D

That model is a bit tricky... the LEDs do not work by default, but I think there was a patch enabling the LEDs which didn't make it past (upstream?) reviewers back then. Other than that, it is solid.

oliv3r commented:
Well the point on buying it would be to make stuff work :)

But thabks for the heads up!

aparcar commented:
Thanks to everyone working on this, sorry if things sometimes take a long time.

To git.openwrt.org:openwrt/openwrt.git
   39b9b491bb5..6af9476b8a3  main -> main

Neustradamus commented:
@aparcar: Thanks for the merging and thanks to @leo5593 this PR and all other too!

plappermaul commented:
Thanks for this one.
